### PR TITLE
Add Dockerfile for building a Docker image running nova

### PR DESCRIPTION
**Pull request by fiadliel**
This Dockerfile should build a Docker image which contains a working "nova" binary.

Can be partnered with a script to link in required directories/env variables at run time.

### CODE REVIEW

**Comment by nataliab** [comment]
Deploy won't work with this script as the nova from the Docker container can't communicate with the Docker on host machine.

**Comment by nataliab** [comment]
It works fine now!

**Comment by gheine** [comment]
Image is missing the nova templates in ~/.nova and - at least when building and running it locally - the aws credentials in ~/.aws

**Comment by fiadliel** [comment]
You have to use the supplied script -- aws credentials have to be linked in to the container, amongst other things.

For the nova templates, it currently uses an env variable to point at your local copy -- I'll change this to bake the templates into the image -- with an override if you want to point at your copy instead.

**Comment by fiadliel** [comment]
Updated:

 - script now requires no env variables, though NOVA_DOCKER_IMAGE can be set to override the name of the image used.
 - clones gilt/nova-shared-templates (at build time), and links to /root/.nova
 - includes a Makefile for building new versions of the Docker image.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+build:
+	docker build -t giltouroboros/nova .
```
Comment (gheine) [inline Makefile]: rename to gilt/nova

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+push:
+	docker push giltouroboros/nova
```
Comment (gheine) [inline Makefile]: same

**Comment by gheine** [comment]
The aws credentials pass-through isn't working for me:

```
IRL-ML-GHEINE1-7:api gheine$ /Users/gheine/Code/nova/nova/bin/nova deploy giltaws staging 1.2.2-40-g7a44304
Traceback (most recent call last):
  File "/usr/local/bin/nova", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python2.7/site-packages/nova/cli/main.py", line 83, in main
    app.run()
  File "/usr/local/lib/python2.7/site-packages/cement/core/foundation.py", line 797, in run
    return_val = self.controller._dispatch()
  File "/usr/local/lib/python2.7/site-packages/cement/core/controller.py", line 472, in _dispatch
    return func()
  File "/usr/local/lib/python2.7/site-packages/cement/core/controller.py", line 478, in _dispatch
    return func()
  File "/usr/local/lib/python2.7/site-packages/nova/cli/controllers/deploy/deploy.py", line 62, in default
    deployer.deploy()
  File "/usr/local/lib/python2.7/site-packages/nova/core/deploy/deploy_stack.py", line 51, in deploy
    print("Deploying to account '%s'..." % self._aws_manager.account_alias)
  File "/usr/local/lib/python2.7/site-packages/nova/core/managers/aws_manager.py", line 45, in account_alias
    self._account_alias = self.iam_client.list_account_aliases()['AccountAliases'][0]
  File "/usr/local/lib/python2.7/site-packages/botocore/client.py", line 278, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/site-packages/botocore/client.py", line 572, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ExpiredToken) when calling the ListAccountAliases operation: The security token included in the request is expired
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/local/lib/python2.7/site-packages/nova/core/deploy/deploy_stack.py", line 260, in __cleanup
    raise NovaError(e)
NovaError: coercing to Unicode: need string or buffer, NoneType found
Error in sys.exitfunc:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/atexit.py", line 24, in _run_exitfuncs
    func(*targs, **kargs)
  File "/usr/local/lib/python2.7/site-packages/nova/core/deploy/deploy_stack.py", line 260, in __cleanup
    raise NovaError(e)
nova.core.exc.NovaError: coercing to Unicode: need string or buffer, NoneType found
```

**Comment by fiadliel** [comment]
botocore.exceptions.ClientError: An error occurred (ExpiredToken) when calling the ListAccountAliases operation: The security token included in the request is expired

You should be better if you run awstoken.

I'll file a bug against nova to give a good error message rather than "coercing to Unicode: need string or buffer, NoneType found"

**Comment by gheine** [comment]
I did and still got that error. But trying the same thing again later it worked... :shrug:

File: Makefile (hunk 1 above)
Comment (fiadliel) [inline Makefile]: Do you have access to the "gilt" Dockerhub account?

**Comment by fiadliel** [comment]
Updated

**Comment by grahamar** [comment]
LGTM 👍  merge when you think it's ready!

**Pull request merged**